core/src/main/java/com/dtolabs/rundeck/core/execution/WorkflowExecutionServiceThread.java

```diff
@@ -79,12 +82,14 @@ public void run() {
 
     public WorkflowExecutionResult runWorkflow() {
         try {
+            iJobPluginService.beforeJobStarts(new JobEventImpl(context));
```

Comment:
the plugin should be able to return a new execution context, and that new context should be used in the `executeWorkflow` on line 87.  When modifying contexts, the pattern is to use the `ExecutionContextImpl.builder(oldContext)` to create a builder from the original context. It can then be modified, and a new one built.  

So in this case, the job plugin service might pass in a builder instead of a StepExecutionContext to the JobPlugin, allowing the plugin to update the context if needed, before returning it to the WorkflowExecutionServiceThread